retrievalmarket/impl/provider.go

```diff
@@ -390,62 +414,180 @@ func (p *Provider) HandleQueryStream(stream rmnet.RetrievalQueryStream) {
+// pieces containing all of the links within that identity CID.
+// Note that it is possible to receive a non-nil error as well as a non-zero length PieceInfo slice
+// as a return from this function. In that case, there was at least one error encountered querying
+// the piece store.
+func (p *Provider) getAllPieceInfoForPayload(payloadCID cid.Cid) ([]piecestore.PieceInfo, error) {
 	// Get all pieces that contain the target block
 	piecesWithTargetBlock, err := p.dagStore.GetPiecesContainingBlock(payloadCID)
 	if err != nil {
-		return piecestore.PieceInfoUndefined, false, xerrors.Errorf("getting pieces for cid %s: %w", payloadCID, err)
+		// this payloadCID may be an identity CID that's in the root of a CAR but
+		// not recorded in the index
+		var idErr error
+		piecesWithTargetBlock, idErr = p.getCommonPiecesFromIdentityCidLinks(payloadCID)
+		if idErr != nil {
+			return []piecestore.PieceInfo{}, idErr
+		}
+		if len(piecesWithTargetBlock) == 0 {
+			return []piecestore.PieceInfo{}, fmt.Errorf("getting pieces for cid %s: %w", payloadCID, err)
+		}
 	}
 
-	// For each piece that contains the target block
+	pieces := make([]piecestore.PieceInfo, 0)
 	var lastErr error
-	var sealedPieceInfo *piecestore.PieceInfo
 	for _, pieceWithTargetBlock := range piecesWithTargetBlock {
 		// Get the deals for the piece
 		pieceInfo, err := p.pieceStore.GetPieceInfo(pieceWithTargetBlock)
 		if err != nil {
 			lastErr = err
 			continue
 		}
+		pieces = append(pieces, pieceInfo)
+	}
 
-		// if client wants to retrieve the payload from a specific piece, just return that piece.
-		if clientPieceCID.Defined() && pieceInfo.PieceCID.Equals(clientPieceCID) {
-			return pieceInfo, p.pieceInUnsealedSector(ctx, pieceInfo), nil
-		}
+	return pieces, lastErr
+}
 
-		// if client doesn't have a preference for a particular piece, prefer a piece
-		// for which an unsealed sector exists.
-		if clientPieceCID.Equals(cid.Undef) {
-			if p.pieceInUnsealedSector(ctx, pieceInfo) {
-				// The piece is in an unsealed sector, so just return it
-				return pieceInfo, true, nil
-			}
+// getCommonPiecesFromIdentityCidLinks will inspect a payloadCID and if it has an identity multihash,
+// will determine which pieces contain all of the links within the decoded identity multihash block
+func (p *Provider) getCommonPiecesFromIdentityCidLinks(payloadCID cid.Cid) ([]cid.Cid, error) {
+	links, err := linksFromIdentityCid(payloadCID)
+	if err != nil || len(links) == 0 {
+		return links, err
+	}
 
-			if sealedPieceInfo == nil {
-				// The piece is not in an unsealed sector, so save it but keep
-				// checking other pieces to see if there is one that is in an
-				// unsealed sector
-				sealedPieceInfo = &pieceInfo
+	pieces := make([]cid.Cid, 0)
+	// for each link, query the dagstore for pieces that contain it
+	for i, link := range links {
+		piecesWithThisCid, err := p.dagStore.GetPiecesContainingBlock(link)
+		if len(piecesWithThisCid) == 0 {
```

Comment:
Is it possible for `GetPiecesContainingBlock` to return a zero-length piecesWithThisCid, and for err to also be nil?

Reply:
Good call, I think it's probably unsafe to not explicitly deal with `err` even though I believe zero-length should also ways be coupled with an error (comes out of `GetShardsForMultihash` in the inverted index which requires an entry for that CID and an ErrNotFound otherwise). Addressed in https://github.com/filecoin-project/go-fil-markets/pull/750